giraffe/src/utils/mosaicTooltip.ts

```diff
@@ -0,0 +1,115 @@
+import {range} from 'd3-array'
+
+import {getRangeLabel} from './tooltip'
+import {X_MIN, X_MAX, FILL, SERIES} from '../constants/columnKeys'
+import {
+  Table,
+  Scale,
+  TooltipData,
+  TooltipColumn,
+  ColumnGroupMap,
+} from '../types'
+
+export const findHoveredBoxes = (
+  boxTable: Table,
+  hoverX: number | null,
+  hoverY: number | null,
+  xScale: Scale<number, number>,
+  yScale: Scale<number, number>,
+  yDomain: number[]
+): number[] => {
+  if (!hoverX || !hoverY) {
+    return []
+  }
+  const xMinData = boxTable.getColumn(X_MIN, 'number')
+  const xMaxData = boxTable.getColumn(X_MAX, 'number')
+  const dataX = xScale.invert(hoverX)
+  const dataY = yScale.invert(hoverY)
+  // Find all boxes whose x extent contain the mouse x position
+  const xIndices = range(0, xMinData.length).filter(
+    i => xMinData[i] <= dataX && xMaxData[i] > dataX
+  )
+
+  const yValMap = new Map()
+  //if series isn't in map yet, add it
+  for (let i = 0; i < yDomain[1]; i++) {
+    const yMin = yScale(i + 1)
+    const yMax = yScale(i)
+    yValMap.set(i, [yMin, yMax])
+  }
+
+  let finalXIndex
+  for (const index of yValMap.keys()) {
+    if (
+      yValMap.get(index)[0] < yScale(dataY) &&
+      yValMap.get(index)[1] >= yScale(dataY)
+    ) {
+      finalXIndex = index
+      break
+    }
+  }
```

Comment:
i'm looking at both of these loops together. a [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) is an object that maintains looping order based on sorting order. Since the items in `yValMap` are being inserted in the order they're looped over, and the main reason to use a `Map` is to maintain order when looping, I think we can combine the functionality of the loops and only loop over these values once.

```ts
for (let i = 0; i < yDomain; i++) {
  const yMin = yScale(i + 1)
  const yMax = yScale(i)

  if (yMin < yScale(dataY) && yMax >= yScale(dataY) {
    return [xIndicies[index]] // if we return early above when xIndicies is falsy this can be changed to a return
  }
}
 // handles the case where the loop didn't return early.
return []
```